lib/rucio/client/uploadclient.py

```diff
@@ -144,11 +144,26 @@ def upload(self, items, summary_file_path=None):
 
             rse = file['rse']
             rse_settings = self.rses[rse]
-            # if file already exists on RSE we're done
-            if rsemgr.exists(rse_settings, file_did):
-                logger.info('File already exists on RSE. Skipping upload')
+            is_deterministic = rse_settings.get('deterministic', True)
+            if not is_deterministic and not pfn:
+                logger.warning('PFN has to be defined for NON-DETERMINISTIC site.')
                 continue
 
+            # if file already exists on RSE we're done
+            if not is_deterministic and not no_register:
+                if rsemgr.exists(rse_settings, pfn):
+                    logger.info('File already exists on RSE with given pfn. Skipping upload. Re-upload is not supported. In such case, existing replica has to be removed first.')
+                    continue
+                elif rsemgr.exists(rse_settings, file_did):
+                    logger.info('File already exists on RSE with different pfn. Skipping upload')
+                    continue
+                else:
```

Comment:
this can be removed

Reply:
it cannot. Threre are three unrelated cases:
1) file exists at site at given pfn
2) file exists somewhere at the site (which you woulldn't catch by first if)
3) file is not at the site and we can upload it.

Reply:
It cannot? What impact does 
else:
  pass

Reply:
ok, you mean only last two lines, ok, they can be removed